src/Microsoft.AspNet.Mvc.Razor/Precompilation/SymbolBasedTypeInfo.cs

```diff
@@ -0,0 +1,99 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using Microsoft.AspNet.Razor.Runtime.TagHelpers;
+using Microsoft.CodeAnalysis;
+
+namespace Microsoft.AspNet.Mvc.Razor.Precompilation
+{
+    [DebuggerDisplay("{FullName}")]
+    public class SymbolBasedTypeInfo : ITypeInfo
+    {
+        private readonly ITypeSymbol _type;
+        private readonly INamedTypeSymbol _openGenericDictionaryTypeSymbol;
+        private string _fullName;
+
+        public SymbolBasedTypeInfo(
+            ITypeSymbol type,
+            INamedTypeSymbol openGenericDictionaryTypeSymbol)
+        {
+            _type = type;
+            _openGenericDictionaryTypeSymbol = openGenericDictionaryTypeSymbol;
+        }
+
+        public string FullName
+        {
+            get
+            {
+                if (_fullName == null)
+                {
+                    _fullName = SymbolUtilities.GetFullName(_type);
+                }
+
+                return _fullName;
+            }
+        }
```

Comment:
It's not necessarily always called - so we just avoid doing extra string concats.